Feature/end to end acceptance tests

nshumoogum opened the pull request:
### What

Added sunny day scenario of the backend and api's as an end to end test to capture the import and export process (hierarchies and search not implemented yet and will be added as separate pieces of work).

Other changes include updating some of the other API tests to use configured mongo databases.

### How to review

Run tests and check they pass:

1) Make sure you have the following environment variables set:
```
export MONGODB_DATABASE=test
export MONGODB_FILTERS_DATABASE=test
export MONGODB_IMPORTS_DATABASE=test
```
Please read README.md for services to be run

2) Checkout the following branches on the following repositories:
* dp-dataset-api - `feature/generate-downloads`
* dp-filter-api - `feature/generate-downloads`
* dp-dataset-exporter-xlsx - `feature/latest-dp-filter`
* dp-dataset-exporter - `feature/generate-downloads`
* All other services should run on cmd-develop but will need to access the environment variables stated above.

3) Run the following test suits and make sure they pass:
* cd filterAPI; go test ./...
* cd datasetAPI; go test ./...
* cd generateFiles; go test ./...

4) Check mongo has not left any documents in test database and not polluted other databases used for manual testing

5) Unset your environment variables:
```
unset MONGODB_DATABASE
unset MONGODB_FILTERS_DATABASE
unset MONGODB_IMPORTS_DATABASE
```

### Who can review

Team B

nshumoogum committed:
Provide the foundations for testing importing and exporting files

Also update vendor package.....it's been a while.

nshumoogum committed:
Include new end to end integration test for backend and apis.

This is still incomplete

nshumoogum committed:
Merge branch 'cmd-develop' into feature/end-to-end-acceptance-tests

nshumoogum committed:
Merge branch 'cmd-develop' into feature/end-to-end-acceptance-tests

nshumoogum committed:
Extend end to end test to include download size and url

This is applied to both filter output resource and version/instance
resource.

nshumoogum committed:
Merge branch 'cmd-develop' into feature/end-to-end-acceptance-tests

nshumoogum committed:
Extend end to end test to check downloadable files

This includes checking the column headers and the number of rows in both
excel and csv files.

eldeal committed:
Merge cmd-develop with feature/end-to-end-acceptance-tests

eldeal committed:
Check edition exists before attempting to access it

eldeal committed:
Merge branch 'cmd-develop' into feature/end-to-end-acceptance-tests

nshumoogum committed:
Check file size is between a small range for e2e tests to pass

nshumoogum committed:
Use configured mongo database for running end to end tests

nshumoogum committed:
Add new configurable variable `MONGO_FILTERS_DATABASE`

Also update end-to-end-environment test to use the correct configured
database for making requests to mongo db.

nshumoogum committed:
Merge branch 'cmd-develop' into feature/end-to-end-acceptance-tests

nshumoogum committed:
Add missing data in filter API test setup

Also use mongo filter database var not generic database var

nshumoogum committed:
Update end to end test README.md

Matt-Guest commented on filterAPI/put_filter_blueprint_test.go:
```diff
@@ -76,37 +75,37 @@ func TestSuccessfulPutFilterBlueprint(t *testing.T) {
-				})
-			})
-		})
+		// Convey("When a request to submit the filter blueprint occurs", func() {
```
why is this test be commented out?

Matt-Guest commented on filterAPI/put_filter_blueprint_test.go:
```diff
@@ -115,58 +114,58 @@ func TestSuccessfulPutFilterBlueprint(t *testing.T) {
-		}
-	})
-}
+// func TestFailureToPutFilterBlueprint(t *testing.T) {
```
Same again here

Matt-Guest commented on generateFiles/end_to_end_api_test.go:
```diff
@@ -0,0 +1,569 @@
+		// Send v4 file to aws
+		_, err := sendV4FileToAWS(region, bucketName, filename)
+		if err != nil {
+			os.Exit(1)
```
Can the error be logged before the tests close

Matt-Guest commented on generateFiles/end_to_end_api_test.go:
```diff
@@ -0,0 +1,569 @@
+				Expect().Status(http.StatusCreated).JSON().Object()
+
+			postJobResponse.Value("id").NotNull()
+			jobID := postJobResponse.Value("id").String().Raw()
```
Why is `.Raw()` used here. Do you just need to call `postJobResponse.Value("id").String()`

nshumoogum replied:
No because jobID should be a string and `postJobResponse.Value("id").String()` is a `*httpexpect.String` type. By using raw function changes the type to string.

Matt-Guest commented on generateFiles/end_to_end_api_test.go:
```diff
@@ -0,0 +1,569 @@
+			postJobResponse.Value("recipe").Equal(recipe)
+			postJobResponse.Value("state").Equal("created")
+
+			instanceID := postJobResponse.Value("links").Object().Value("instances").Array().Element(0).Object().Value("id").String().Raw()
```
Same again here with the `Raw()` function

nshumoogum replied:
Same reason as above

Matt-Guest commented on generateFiles/end_to_end_api_test.go:
```diff
@@ -0,0 +1,569 @@
+			totalObservations := 1513
+
+			tryAgain := true
+			for tryAgain {
```
Should there be a timeout here? As this could lead to an infinite loop

Matt-Guest commented on generateFiles/end_to_end_api_test.go:
```diff
@@ -0,0 +1,569 @@
+			// Waiting for version to have downloads before updating state to published
+			hasDownloads := false
+			var XLSSize int
+			for !hasDownloads {
```
Should we have a time limit for this for loop?

Matt-Guest commented on generateFiles/end_to_end_api_test.go:
```diff
@@ -0,0 +1,569 @@
+					filteredCSVReader := csv.NewReader(filteredCSVFile)
+
+					if err = checkFileRowCount(filteredCSVReader, filteredCSVS3URL, 39); err != nil {
+						os.Exit(1)
```
The error should be logged before exiting

Matt-Guest commented on generateFiles/setup.go:
```diff
@@ -0,0 +1,197 @@
+		return nil, err
+	}
+
+	file, err := s3.Get(s3URL)
```
Could you just use `return s3.Get(s3URL)`? and remove the lines below?

Matt-Guest commented on testDataSetup/mongo/mongo.go:
```diff
@@ -89,6 +106,41 @@ func Setup(d ...*Doc) error {
+
+// UploadedFile used for a file which has been uploaded to a bucket
+type UploadedFile struct {
+	AliasName string `bson:"alias_name" json:"alias_name" avro:"alias-name"`
```
Do you need the avro tags?

Matt-Guest commented on testDataSetup/mongo/mongo.go:
```diff
@@ -339,6 +391,20 @@ func GetInstance(database, collection, key, value string) (Instance, error) {
+	defer s.Close()
+
+	var count int
+	count, err := s.DB(database).C(collection).Find(bson.M{key: value}).Count()
```
Can this be changed to `return s.DB(database).C(collection).Find(bson.M{key: value}).Count()`, as err will either be nil or an error type.

nshumoogum committed:
Cleanup test database on initialisation of e2e test

Also:
*Uncomment tests that were previously accidently commented out
*Added extra error log lines prior to exiting test
*Remove avro tags from field definition in structs
*Exit out of infinite loops after 30 seconds

Matt-Guest reviewed (approved):
Do you mind renaming the directory name for `generateFiles`?

nshumoogum committed:
Rename directory generateFiles to endTpEndTests

The test coverage within directory, tests a sunny scenario across all
backend services (including API's) and hence does not just test the
generation of files eventhough this was the cause of writing this e2e
test suite.

nshumoogum committed:
Make timeout a global variable for reuse in tests
